### ISSUE TITLE
Truncated output segments

### ISSUE THREAD

**Issue by aretius** [microsoft/unilm#72]
Hey all,

I have been facing an issue with UniLM V1, where the model has been fine-tuned for a Case correction task. Case correction is the problem where the model figures out which words need to be capitalized. The input to the model is all words in Capital case(first character of word Capital) and output being the correct case.
However, the predictions have been truncated. I have used the model within the summarization & questions generation task; however, I observed truncation when the `max_seq_length` or `max_pred` were set quite low. Here the max number of tokens in the input & output is 70.

 I used the following snippet for fine-tuning and predictions, respectively.

FineTuning
```
python biunilm/run_seq2seq.py --fp16 --amp --do_train --num_workers 0 \
--bert_model bert-large-cased --new_segment_ids \
--data_dir ${DATA_DIR} --src_file train.src.txt --tgt_file train.tgt.txt \
--output_dir ${OUTPUT_DIR}/bert_save \
  --log_dir ${OUTPUT_DIR}/bert_log \
  --model_recover_path ${MODEL_RECOVER_PATH} \
  --max_seq_length 200 --max_position_embeddings 200 \
  --mask_prob 0.7 --max_pred 150 \
  --train_batch_size 128 --gradient_accumulation_steps 6 \
  --learning_rate 0.00002 --warmup_proportion 0.1 --label_smoothing 0.1 \
  --num_train_epochs 10

```
Prediction
```
python biunilm/decode_seq2seq.py --fp16 --amp --bert_model bert-large-cased \
--new_segment_ids --mode s2s \
  --input_file ${DATA_DIR}/${EVAL_SPLIT}.src.txt --split ${EVAL_SPLIT} \
  --model_recover_path ${MODEL_RECOVER_PATH} \
  --max_seq_length 200 --max_tgt_length 150 \
  --batch_size 128 --beam_size 5 --length_penalty 0 \
  --not_predict_token "[UNK]" --forbid_duplicate_ngrams --forbid_ignore_word "."
```

However, I am observing about 10% of cases where the output is truncated like

Model output:
> Masked against coronavirus, Shanghai's ballet dancers pirouette and 

Correct output

> Masked against coronavirus, Shanghai's ballet dancers pirouette and plie

Any help will be appreciated, thanks!

**Comment by donglixp** [comment]
Hi @aretius ,

We haven't encountered the issue before. In the training data, whether there were some examples being truncated? For the case correction task, because we can know the alignments between input and output, we can constrain the candidate predictions and length. The model can be forced to predict the desired number of tokens. Another solution would be trying the new package https://github.com/microsoft/unilm/tree/master/s2s-ft released for s2s fine-tuning.

**Comment by aretius** [comment]
Thanks for the response @donglixp !
Yes, it is indeed quite weird. I am also attaching a distribution of token length in input.
![pastedImagebase640](https://user-images.githubusercontent.com/18247856/76219617-1d752e00-623c-11ea-91dd-e354793753d1.png)
Just had a few follow-up questions

- Is it possible that the model is biased to produce outputs of the average token length of input(since the dataset size is ~8K data points)?
- Since the input has all words cased the # of tokens in input and correct output are different. So I am not sure how will we end up constraining the candidate predictions and length. Could you help me with the code piece in the repo that will allow me to do the same?
- How is the s2s-ft different from UniLM v1 code piece?

**Comment by aretius** [comment]
@donglixp I observed that training with a larger dataset reduces the truncation rate to a mere 2%. When having a smaller dataset, can you provide some info on the epochs to run it for?

**Comment by donglixp** [comment]
@aretius For the case correction task, because we can know whether the output is completely decoded, we can forbid the end-of-sequence token until all the desired tokens are emitted. For example, if we found that the current prediction is `Masked against coronavirus, Shanghai's ballet dancers pirouette and` (i.e., the word `plie` is missing), we could assign a large negative score for <EOS>, so that the output won't be truncated.
